### PR TITLE
CATO-3540: Adding validation for 'E30' and 'E35'.

### PR DESCRIPTION
**Pull request by tmikus**

### CODE REVIEW

File: src/main/scala/uk/gov/hmrc/ct/ct600e/v3/E30.scala [hunk 1]
```diff
@@ -20,5 +20,6 @@ import uk.gov.hmrc.ct.box._
 import uk.gov.hmrc.ct.ct600e.v3.retriever.CT600EBoxRetriever
 
 case class E30(value: Option[String]) extends CtBoxIdentifier("Claimer's name") with CtOptionalString with Input with ValidatableBox[CT600EBoxRetriever]{
-  override def validate(boxRetriever: CT600EBoxRetriever): Set[CtValidation] = validateStringAsMandatory("E30", this)
+  override def validate(boxRetriever: CT600EBoxRetriever): Set[CtValidation] =
+    validateStringAsMandatory("E30", this) ++ validateOptionalStringByLength("E30", this, 2, 56) ++ validateOptionalStringByRegex("E30", this, validNonForeignLessRestrictiveCharacters)
```
Comment (liquidarmour) [inline src/main/scala/uk/gov/hmrc/ct/ct600e/v3/E30.scala]: Looks like almost a copy of below. Can we share this logic somehow?

File: src/main/scala/uk/gov/hmrc/ct/box/ValidatableBox.scala [hunk 2]
```diff
@@ -227,4 +227,7 @@ trait ValidatableBox[T <: BoxRetriever] {
     }
   }
 
+  def validateDeclarationNameOrStatus(boxId: String, box: CtOptionalString): Set[CtValidation] = {
```
Comment (scottcutts) [inline src/main/scala/uk/gov/hmrc/ct/box/ValidatableBox.scala]: I would pull this out into it's own trait as its specific validation and not generic, like the Registered Charity Number

File: src/test/scala/uk/gov/hmrc/ct/ct600e/v3/E30Spec.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+package uk.gov.hmrc.ct.ct600e.v3
+
+import org.mockito.Mockito._
+import org.scalatest.mock.MockitoSugar
+import org.scalatest.{Matchers, WordSpec}
+import uk.gov.hmrc.ct.box.CtValidation
+import uk.gov.hmrc.ct.ct600e.v3.retriever.CT600EBoxRetriever
+
+class E30Spec extends WordSpec with MockitoSugar with Matchers {
```
Comment (scottcutts) [inline src/test/scala/uk/gov/hmrc/ct/ct600e/v3/E30Spec.scala]: Looks like the same tests twice. You can just test the trait once instead.

File: src/test/scala/uk/gov/hmrc/ct/ct600e/validations/ValidateDeclarationNameOrStatusSpec.scala [hunk 4]
```diff
@@ -14,54 +14,67 @@
...
           val value: Option[String] = Some("123456789 123456789 123456789 123456789 123456789 1234567")
           val mockBoxRetriever = mock[CT600EBoxRetriever]
-          when(mockBoxRetriever.retrieveE1030()).thenReturn(E1030(value))
-      
-          E1030(value).validate(mockBoxRetriever) shouldBe Set(CtValidation(Some("E1030"), "error.E1030.text.sizeRange", Some(Seq("2", "56"))))
+          val box = E30(value)
+
+          when(mockBoxRetriever.retrieveE30()).thenReturn(box)
+
+          validateDeclarationNameOrStatus("E30", box) shouldBe Set(CtValidation(Some("E30"), "error.E30.text.sizeRange", Some(Seq("2", "56"))))
```
Comment (liquidarmour) [inline src/test/scala/uk/gov/hmrc/ct/ct600e/validations/ValidateDeclarationNameOrStatusSpec.scala]: Does this work with a different box number?

Comment (tmikus) [inline src/test/scala/uk/gov/hmrc/ct/ct600e/validations/ValidateDeclarationNameOrStatusSpec.scala]: It's a trait which depends on other methods. Those methods REQUIRE a box ID and box so unfortunately I had to write a test for a specific box.

Comment (tmikus) [inline src/test/scala/uk/gov/hmrc/ct/ct600e/validations/ValidateDeclarationNameOrStatusSpec.scala]: I could re-implement those validation methods in generic way but then that would require refactoring of half of ct-calculations.

**Pull request merged**